include/envoy/http/header_map.h

```diff
@@ -601,6 +627,16 @@ class HeaderMap {
    */
   virtual void dumpState(std::ostream& os, int indent_level = 0) const PURE;
 
+  /**
+   * @return true if header field-name case-preservation is enabled in map, false otherwise.
+   */
+  virtual bool isHeadersCasePreservationEnabled() const PURE;
```

Comment:
I prefer `bool headerKeysCasePreserved()` if we have to have this. But I wonder if it really needs to be a boolean at a map level, or if instead it's per HeaderEntry (i.e. some cases are preserved and some aren't, so the config value only affects incoming headers)

Reply:
I think we have to have this. Because we are planning to have wrapper functions such as for existing `void addCopy(const LowerCaseString& key, absl::string_view value);`, we are planning to define a wrapper such as:

```
void addCopy(const absl::string_view key, absl::string_view value) {
  this->addCopy(LowerCaseString(key), value);
  if (headerKeysCasePreserved()) {
    addCasePreservedCopy(LowerCaseString(key), key);
  }
}

```
The purpose of these wrapper functions is to give the filter-developer an option to use case-preserving `addCopy` if they want to preserve the header case.
Now, this new` addCopy `will be part of the `HeaderMapImpl`. This headerMapImpl object does not have access to codec_settings (`codec_settings_.header_key_format_` i.e. `Http1Settings::HeaderKeyFormat::PreservedCase`).
To make `addCopy` take decision-based on whether preserved-casing is set or not, we need to use a function like `setHeaderKeysCasePreserved(true)` in the header-map to make the header map know about it, at the very beginning when the first incoming header arrives in the codec.

Also, I think that boolean have to be at the map level. The only purpose is to make header-map know that codec_settings are set to preserve the header case for incoming headers. So that when a new header-entry is created using `addCopy`, the header-map knows whether to populate the header-entry with `case_preserved_key` or not.

I agree that some headers keys are preserved (incoming headers), some are not preserved (programmatically created headers). Thus, the name of the function can be
```
bool incomingHeaderKeysCasePreserved();
void setIncomingHeaderKeysCasePreserved(bool enabled);

```
I hope it make sense?

Reply:
Even if we don't have this wrapper addCopy(), then everywhere we want case preservation, we need to call addCasePreservedCopy() based on if the PreservedCase is set or not in codec_settings which may now be accessible and that map level boolean may be the only way to take decision.